File: python/data_structures/linked_list.py
"""
Singly Linked List Implementation
Time Complexity:
- Insertion: O(1) at head, O(n) at tail or specific position
- Searching: O(n)
- Deletion: O(n)
- Traversal: O(n)
"""

class Node:
    """Node class for singly linked list"""
    def __init__(self, data):
        self.data = data
        self.next = None

class SinglyLinkedList:
    """Singly Linked List with common operations"""
    
    def __init__(self):
        self.head = None
    
    # 2.1.1 Insertion
    def insert_at_beginning(self, data):
        """Insert node at the beginning - O(1)"""
        new_node = Node(data)
        new_node.next = self.head
        self.head = new_node
    
    def insert_at_end(self, data):
        """Insert node at the end - O(n)"""
        new_node = Node(data)
        
        if not self.head:
            self.head = new_node
            return
        
        current = self.head
        while current.next:
            current = current.next
        current.next = new_node
    
    def insert_at_position(self, data, position):
        """Insert node at specific position - O(n)"""
        if position == 0:
            self.insert_at_beginning(data)
            return
        
        new_node = Node(data)
        current = self.head
        
        for i in range(position - 1):
            if not current:
                print("Position out of bounds")
                return
            current = current.next
        
        new_node.next = current.next
        current.next = new_node
    
    # 2.1.2 Searching
    def search(self, key):
        """Search for a node with given key - O(n)"""
        current = self.head
        position = 0
        
        while current:
            if current.data == key:
                return position
            current = current.next
            position += 1
        
        return -1  # Not found
    
    def contains(self, key):
        """Check if list contains a key - O(n)"""
        return self.search(key) != -1
    
    # 2.1.3 Deletion
    def delete_by_value(self, key):
        """Delete first node with given value - O(n)"""
        if not self.head:
            return
        
        # If head needs to be deleted
        if self.head.data == key:
            self.head = self.head.next
            return
        
        current = self.head
        while current.next:
            if current.next.data == key:
                current.next = current.next.next
                return
            current = current.next
    
    def delete_at_position(self, position):
        """Delete node at specific position - O(n)"""
        if not self.head:
            return
        
        if position == 0:
            self.head = self.head.next
            return
        
        current = self.head
        for i in range(position - 1):
            if not current.next:
                print("Position out of bounds")
                return
            current = current.next
        
        if current.next:
            current.next = current.next.next
    
    # 2.1.4 Traversing the list
    def traverse(self):
        """Traverse and print all elements - O(n)"""
        elements = []
        current = self.head
        
        while current:
            elements.append(current.data)
            current = current.next
        
        return elements
    
    def display(self):
        """Display the linked list"""
        elements = self.traverse()
        print(" -> ".join(map(str, elements)) + " -> None")
    
    # 2.1.5 Traversing the list in reverse order
    def reverse_traverse_recursive(self, node=None):
        """Traverse in reverse using recursion - O(n)"""
        if node is None:
            node = self.head
        
        if node is None:
            return []
        
        if node.next is None:
            return [node.data]
        
        return self.reverse_traverse_recursive(node.next) + [node.data]
    
    def reverse(self):
        """Reverse the linked list - O(n)"""
        prev = None
        current = self.head
        
        while current:
            next_node = current.next
            current.next = prev
            prev = current
            current = next_node
        
        self.head = prev
    
    def get_length(self):
        """Get the length of the list - O(n)"""
        count = 0
        current = self.head
        
        while current:
            count += 1
            current = current.next
        
        return count
    
    def is_empty(self):
        """Check if list is empty - O(1)"""
        return self.head is None


# Example usage and testing
if __name__ == "__main__":
    print("=== Singly Linked List Demo ===\n")
    
    # Create a new linked list
    sll = SinglyLinkedList()
    
    # Insertion operations
    print("1. Insertion Operations:")
    sll.insert_at_beginning(3)
    sll.insert_at_beginning(2)
    sll.insert_at_beginning(1)
    print("After inserting 1, 2, 3 at beginning:")
    sll.display()
    
    sll.insert_at_end(4)
    sll.insert_at_end(5)
    print("After inserting 4, 5 at end:")
    sll.display()
    
    sll.insert_at_position(2.5, 2)
    print("After inserting 2.5 at position 2:")
    sll.display()
    
    # Searching operations
    print("\n2. Searching Operations:")
    print(f"Search for 3: Found at position {sll.search(3)}")
    print(f"Search for 10: Found at position {sll.search(10)}")
    print(f"Contains 4: {sll.contains(4)}")
    print(f"Contains 100: {sll.contains(100)}")
    
    # Deletion operations
    print("\n3. Deletion Operations:")
    sll.delete_by_value(2.5)
    print("After deleting value 2.5:")
    sll.display()
    
    sll.delete_at_position(0)
    print("After deleting at position 0:")
    sll.display()
    
    # Traversal operations
    print("\n4. Traversal Operations:")
    print(f"Forward traversal: {sll.traverse()}")
    print(f"Reverse traversal: {sll.reverse_traverse_recursive()}")
    
    # Reverse the list
    print("\n5. Reversing the list:")
    sll.reverse()
    print("After reversing:")
    sll.display()
    
    # Other operations
    print("\n6. Other Operations:")
    print(f"Length of list: {sll.get_length()}")
    print(f"Is empty: {sll.is_empty()}")

File: c/data_structures/linked_list.c
/*
 * Singly Linked List Implementation in C
 * Time Complexity:
 * - Insertion: O(1) at head, O(n) at tail
 * - Searching: O(n)
 * - Deletion: O(n)
 * - Traversal: O(n)
 */

#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>

// Node structure
typedef struct Node {
    int data;
    struct Node* next;
} Node;

// Create a new node
Node* create_node(int data) {
    Node* new_node = (Node*)malloc(sizeof(Node));
    if (new_node == NULL) {
        printf("Memory allocation failed\n");
        exit(1);
    }
    new_node->data = data;
    new_node->next = NULL;
    return new_node;
}

// Insert at beginning - O(1)
Node* insert_at_beginning(Node* head, int data) {
    Node* new_node = create_node(data);
    new_node->next = head;
    return new_node;
}

// Insert at end - O(n)
Node* insert_at_end(Node* head, int data) {
    Node* new_node = create_node(data);
    
    if (head == NULL) {
        return new_node;
    }
    
    Node* current = head;
    while (current->next != NULL) {
        current = current->next;
    }
    current->next = new_node;
    
    return head;
}

// Insert at position - O(n)
Node* insert_at_position(Node* head, int data, int position) {
    if (position == 0) {
        return insert_at_beginning(head, data);
    }
    
    Node* new_node = create_node(data);
    Node* current = head;
    
    for (int i = 0; i < position - 1 && current != NULL; i++) {
        current = current->next;
    }
    
    if (current == NULL) {
        printf("Position out of bounds\n");
        free(new_node);
        return head;
    }
    
    new_node->next = current->next;
    current->next = new_node;
    
    return head;
}

// Search for a value - O(n)
int search(Node* head, int key) {
    Node* current = head;
    int position = 0;
    
    while (current != NULL) {
        if (current->data == key) {
            return position;
        }
        current = current->next;
        position++;
    }
    
    return -1; // Not found
}

// Delete by value - O(n)
Node* delete_by_value(Node* head, int key) {
    if (head == NULL) {
        return NULL;
    }
    
    // If head needs to be deleted
    if (head->data == key) {
        Node* temp = head;
        head = head->next;
        free(temp);
        return head;
    }
    
    Node* current = head;
    while (current->next != NULL) {
        if (current->next->data == key) {
            Node* temp = current->next;
            current->next = current->next->next;
            free(temp);
            return head;
        }
        current = current->next;
    }
    
    return head;
}

// Delete at position - O(n)
Node* delete_at_position(Node* head, int position) {
    if (head == NULL) {
        return NULL;
    }
    
    if (position == 0) {
        Node* temp = head;
        head = head->next;
        free(temp);
        return head;
    }
    
    Node* current = head;
    for (int i = 0; i < position - 1 && current->next != NULL; i++) {
        current = current->next;
    }
    
    if (current->next == NULL) {
        printf("Position out of bounds\n");
        return head;
    }
    
    Node* temp = current->next;
    current->next = current->next->next;
    free(temp);
    
    return head;
}

// Traverse and display - O(n)
void display(Node* head) {
    Node* current = head;
    
    while (current != NULL) {
        printf("%d -> ", current->data);
        current = current->next;
    }
    printf("NULL\n");
}

// Reverse the list - O(n)
Node* reverse(Node* head) {
    Node* prev = NULL;
    Node* current = head;
    Node* next = NULL;
    
    while (current != NULL) {
        next = current->next;
        current->next = prev;
        prev = current;
        current = next;
    }
    
    return prev;
}

// Get length - O(n)
int get_length(Node* head) {
    int count = 0;
    Node* current = head;
    
    while (current != NULL) {
        count++;
        current = current->next;
    }
    
    return count;
}

// Free all nodes - O(n)
void free_list(Node* head) {
    Node* current = head;
    Node* next;
    
    while (current != NULL) {
        next = current->next;
        free(current);
        current = next;
    }
}

// Main function for testing
int main() {
    printf("=== Singly Linked List Demo (C) ===\n\n");
    
    Node* head = NULL;
    
    // Insertion operations
    printf("1. Insertion Operations:\n");
    head = insert_at_beginning(head, 3);
    head = insert_at_beginning(head, 2);
    head = insert_at_beginning(head, 1);
    printf("After inserting 1, 2, 3 at beginning: ");
    display(head);
    
    head = insert_at_end(head, 4);
    head = insert_at_end(head, 5);
    printf("After inserting 4, 5 at end: ");
    display(head);
    
    head = insert_at_position(head, 10, 2);
    printf("After inserting 10 at position 2: ");
    display(head);
    
    // Searching operations
    printf("\n2. Searching Operations:\n");
    printf("Search for 3: Found at position %d\n", search(head, 3));
    printf("Search for 100: Found at position %d\n", search(head, 100));
    
    // Deletion operations
    printf("\n3. Deletion Operations:\n");
    head = delete_by_value(head, 10);
    printf("After deleting value 10: ");
    display(head);
    
    head = delete_at_position(head, 0);
    printf("After deleting at position 0: ");
    display(head);
    
    // Other operations
    printf("\n4. Other Operations:\n");
    printf("Length of list: %d\n", get_length(head));
    
    printf("\n5. Reversing the list:\n");
    head = reverse(head);
    printf("After reversing: ");
    display(head);
    
    // Clean up
    free_list(head);
    
    return 0;
}
